src/closure.rs

```diff
@@ -0,0 +1,127 @@
+// TODO: support marshaller.
+
+use std::mem;
+use std::ptr;
+
+use libc::{c_uint, c_void};
+
+use gobject_ffi;
+use translate::{FromGlibPtrFull, FromGlibPtrNone, Stash, StashMut, ToGlibPtr, ToGlibPtrMut, Uninitialized};
+use Value;
+use ToValue;
+
+pub struct Closure {
+    closure: *mut gobject_ffi::GClosure,
+}
+
+impl Closure {
+    pub fn new<F: Fn(&[Value]) -> Option<Value> + 'static>(callback: F) -> Self {
+        unsafe extern "C" fn marshal<F: Fn(&[Value]) -> Option<Value> + 'static>(closure: *mut gobject_ffi::GClosure,
+            return_value: *mut gobject_ffi::GValue, n_param_values: c_uint, mut param_values: *mut gobject_ffi::GValue,
+            invocation_hint: *mut c_void, marshal_data: *mut c_void)
+        {
+            let mut values = vec![];
+            for _ in 0 .. n_param_values {
+                let value: Value = FromGlibPtrNone::from_glib_none(param_values);
+                values.push(value);
+                param_values = param_values.offset(1);
+            }
+            let callback: Box<Box<F>> = Box::from_raw(marshal_data as *mut _);
+            let result = callback(&values);
+            if let Some(result) = result {
+                *return_value = result.into_raw();
+            }
+            for value in values {
+                mem::forget(value);
```

Comment:
And this is why the FromGlib* impls work for you. Once fixed, you would have to remove the mem::forget() here and let the Drop impls take care of everything else